dll/win32/shell32/folders/CDesktopFolder.cpp

```diff
@@ -447,22 +446,67 @@ HRESULT WINAPI CDesktopFolder::BindToStorage(
 HRESULT WINAPI CDesktopFolder::CompareIDs(LPARAM lParam, PCUIDLIST_RELATIVE pidl1, PCUIDLIST_RELATIVE pidl2)
 {
     bool bIsDesktopFolder1, bIsDesktopFolder2;
+    UINT iColumnModified = LOWORD(lParam);
+
+    if (LOWORD(lParam) >= DESKTOPSHELLVIEWCOLUMNS)
+    {
+        ERR("Got invalid column pointer %x\n", LOWORD(lParam));
+        return E_INVALIDARG;
+    }
 
     if (!pidl1 || !pidl2)
     {
-        ERR("Got null pidl pointer (%Ix %p %p)!\n", lParam, pidl1, pidl2);
+        ERR("Got null pidl pointer (%x %p %p)!\n", LOWORD(lParam), pidl1, pidl2);
         return E_INVALIDARG;
     }
 
+    if (DesktopSFHeader[LOWORD(lParam)].colnameid == IDS_SHV_COLUMN_MODIFIED)
+    {
+        LPPIDLDATA pData1 = _ILGetDataPointer(pidl1);
+        LPPIDLDATA pData2 = _ILGetDataPointer(pidl2);
+        BOOL bIsSpecialFolder1 = _ILIsSpecialFolder(pidl1);
+        BOOL bIsSpecialFolder2 = _ILIsSpecialFolder(pidl2);
+        int result;
+
+        /* If both are Special Folders then it does not matter which is first */
+        if (bIsSpecialFolder1 && bIsSpecialFolder2)
+            return MAKE_COMPARE_HRESULT(0);
+
+        /* If only one is a Special Folder, then the Special Folder is first */
+        if (bIsSpecialFolder1 || bIsSpecialFolder2)
+            return MAKE_COMPARE_HRESULT(bIsSpecialFolder1 ? -1 : 1);
+
+        /* Otherwise, compare on date/time */
+        result = pData1->u.file.uFileDate - pData2->u.file.uFileDate;
+        if (result == 0)
+        {
+            result = pData1->u.file.uFileTime - pData2->u.file.uFileTime;
+            if (result == 0)
+                return SHELL32_CompareChildren(this, LOWORD(lParam), pidl1, pidl2);
```

Comment:
LOWORD here is incorrect, when the lParam ends up in CompareIDs then it needs the full 32-bits because there can be flags there. (Don't ask me why it is LPARAM and not UINT).

Reply:
Fixed. Thanks.